### ISSUE TITLE
Error when compressing files of specific size with -B4 on the command line

### ISSUE THREAD

**Issue by MarcMassar** [lz4/lz4#1374]
**Describe the bug**

Error when compressing files of a size 65533 to 65535 bytes when using "lz4 -B4"

**To Reproduce**
On the command line:

[lz4_err]$ yes | head -c 65533 | lz4 -B4 >file_cout
using blocks of size 64 KB 
Error 41 : Compression failed : ERROR_dstMaxSize_tooSmall 

[lz4_err]$ echo $?
41

[lz4_err]$ ls -l
total 0
-rw-r--r--. 1 marcmassar marcmassar 0 Mar  6 22:28 file_cout

**System (please complete the following information):**
 - OS: tested on rel8, fedora. 
 
[lz4_err]$ lz4 --version
*** LZ4 command line interface 64-bits v1.9.4, by Yann Collet ***

**Comment by t-mat** [comment]
Hi @MarcMassar 
Thank you for the report.

First of all, I've confirmed your issue with `v1.9.4`.  I believe it's our bug.
But I also don't confirm this issue with the recent (not-released) version of lz4 repo.
It means that this bug had been (accidentally) fixed.

Therefore, if you need the fix now, please pull the recent version lz4 from `dev` branch of this repo.
Actual commit hash of the fix is e58a06d1f9df02c85bfa15abcfd97118f9d355d6 (part of #1336).

----

We can find the first *good* commit with [this method by Michael Wolf](https://stackoverflow.com/a/17153598).

<details><summary>git bisect log</summary>

```
/lz4-issue-1374$ git bisect start --term-new=fixed --term-old=unfixed
/lz4-issue-1374$ git bisect fixed dev
/lz4-issue-1374$ git bisect unfixed 5ff839680134437dbf4678f3d0c7b371d84f4964


Bisecting: a merge base must be tested
[cfd6ab32522280079c2e6d3ea995f172b9ae0312] update NEWS for v1.9.4
/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
cfd6ab3 (HEAD) update NEWS for v1.9.4
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
make[1]: Leaving directory '/lz4-issue-1374/programs'
using blocks of size 64 KB
Error 41 : Compression failed : ERROR_dstMaxSize_tooSmall
/lz4-issue-1374$ git bisect unfixed
Bisecting: 230 revisions left to test after this (roughly 8 steps)
[2b755f4594b7e9bf8247394c3a2474621b138412] Merge pull request #1302 from lz4/dependabot/github_actions/github/codeql-action-2.22.4


/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
2b755f4 (HEAD) Merge pull request #1302 from lz4/dependabot/github_actions/github/codeql-action-2.22.4
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
make[1]: Leaving directory '/lz4-issue-1374/programs'
using blocks of size 64 KB
Error 41 : Compression failed : ERROR_dstMaxSize_tooSmall
/lz4-issue-1374$ git bisect unfixed
Bisecting: 115 revisions left to test after this (roughly 7 steps)
[db7495d08b52ef9e51c4233a9b3c82d0739d2ec0] fix another C90 pedantic warning


/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
db7495d (HEAD) fix another C90 pedantic warning
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
==> building with multithreading support
make[1]: Leaving directory '/lz4-issue-1374/programs'
lz4 build completed
using blocks of size 64 KB
/lz4-issue-1374$ git bisect fixed
Bisecting: 57 revisions left to test after this (roughly 6 steps)
[5c9e349c4a7ec8f6449f37f26e24882586b35a82] changed test name cpp->cxx


/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
5c9e349 (HEAD) changed test name cpp->cxx
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
make[1]: Leaving directory '/lz4-issue-1374/programs'
using blocks of size 64 KB
Error 41 : Compression failed : ERROR_dstMaxSize_tooSmall
/lz4-issue-1374$ git bisect unfixed
Bisecting: 28 revisions left to test after this (roughly 5 steps)
[67648edb8c5aba25aa94836215f51e15225e8b81] Merge pull request #1331 from lz4/man_mt


/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
67648ed (HEAD) Merge pull request #1331 from lz4/man_mt
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
make[1]: Leaving directory '/lz4-issue-1374/programs'
using blocks of size 64 KB
Error 41 : Compression failed : ERROR_dstMaxSize_tooSmall
/lz4-issue-1374$ git bisect unfixed
Bisecting: 14 revisions left to test after this (roughly 4 steps)
[e58a06d1f9df02c85bfa15abcfd97118f9d355d6] fixed smaller block sizes


/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
e58a06d (HEAD) fixed smaller block sizes
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
util.c: In function ‘UTIL_countCores’:
util.c:175:13: warning: variable ‘ratio’ set but not used [-Wunused-but-set-variable]
  175 |         int ratio = 1;
      |             ^~~~~
make[1]: Leaving directory '/lz4-issue-1374/programs'
using blocks of size 64 KB
/lz4-issue-1374$ git bisect fixed
Bisecting: 6 revisions left to test after this (roughly 3 steps)
[127fa5789502e112e2824c004168c546a45bf401] control over nbThreads


/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
127fa57 (HEAD) control over nbThreads
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
lz4io.c: In function ‘LZ4IO_setNbWorkers’:
lz4io.c:199:1: warning: control reaches end of non-void function [-Wreturn-type]
  199 | }
      | ^
util.c: In function ‘UTIL_countCores’:
util.c:175:13: warning: variable ‘ratio’ set but not used [-Wunused-but-set-variable]
  175 |         int ratio = 1;
      |             ^~~~~
make[1]: Leaving directory '/lz4-issue-1374/programs'
using blocks of size 64 KB
Error 41 : Compression failed : ERROR_dstMaxSize_tooSmall
/lz4-issue-1374$ git bisect unfixed
Bisecting: 3 revisions left to test after this (roughly 2 steps)
[3acabda180389bfc5ea095c6ffcb1e71562ea7ea] multithreading works with lz4f format, but


/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
3acabda (HEAD) multithreading works with lz4f format, but
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
util.c: In function ‘UTIL_countCores’:
util.c:175:13: warning: variable ‘ratio’ set but not used [-Wunused-but-set-variable]
  175 |         int ratio = 1;
      |             ^~~~~
make[1]: Leaving directory '/lz4-issue-1374/programs'
using blocks of size 64 KB
Error 41 : Compression failed : ERROR_dstMaxSize_tooSmall
/lz4-issue-1374$ git bisect unfixed
Bisecting: 1 revision left to test after this (roughly 1 step)
[d4c21a4dd0aafee521d4da0a2ebadd17a427b200] fixed stdin support


/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
d4c21a4 (HEAD) fixed stdin support
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
util.c: In function ‘UTIL_countCores’:
util.c:175:13: warning: variable ‘ratio’ set but not used [-Wunused-but-set-variable]
  175 |         int ratio = 1;
      |             ^~~~~
make[1]: Leaving directory '/lz4-issue-1374/programs'
using blocks of size 64 KB
Error 41 : Compression failed : ERROR_dstMaxSize_tooSmall
/lz4-issue-1374$ git bisect unfixed
Bisecting: 0 revisions left to test after this (roughly 0 steps)
[95c2bc7f1f6dac251d95853cde461f55e9562a16] fix legacy format compression


/lz4-issue-1374$ (git log -1 --oneline) && (make clean lz4) && (yes | head -c 65533 | ./lz4 -B4 >file_cout)
95c2bc7 (HEAD) fix legacy format compression
Cleaning completed
make[1]: Entering directory '/lz4-issue-1374/lib'
compiling static library
make[1]: Leaving directory '/lz4-issue-1374/lib'
make[1]: Entering directory '/lz4-issue-1374/programs'
util.c: In function ‘UTIL_countCores’:
util.c:175:13: warning: variable ‘ratio’ set but not used [-Wunused-but-set-variable]
  175 |         int ratio = 1;
      |             ^~~~~
make[1]: Leaving directory '/lz4-issue-1374/programs'
using blocks of size 64 KB
Error 41 : Compression failed : ERROR_dstMaxSize_tooSmall
/lz4-issue-1374$ git bisect unfixed
e58a06d1f9df02c85bfa15abcfd97118f9d355d6 is the first fixed commit
commit e58a06d1f9df02c85bfa15abcfd97118f9d355d6
Author: Yann Collet <cyan@fb.com>
Date:   Wed Dec 27 15:59:53 2023 -0800

    fixed smaller block sizes

 lib/lz4frame.c   |  2 +-
 programs/lz4io.c | 71 +++++++++++++++++++++++++++++++++-----------------------
 2 files changed, 43 insertions(+), 30 deletions(-)
```

</details>


<details><summary>Confirmation</summary>

```
cd
git clone https://github.com/lz4/lz4.git lz4-issue-1374-check
cd lz4-issue-1374-check

git log -3 --pretty=format:"%h %ad %s" e58a06d
# e58a06d Wed Dec 27 15:59:53 2023 -0800 fixed smaller block sizes
# 95c2bc7 Wed Dec 27 15:22:53 2023 -0800 fix legacy format compression
# d4c21a4 Wed Dec 27 14:17:02 2023 -0800 fixed stdin support

git checkout e58a06d
git log -1 && make && ./lz4 --version && ( yes | head -c 65533 | ./lz4 -B4 >file_cout )
# OK

git checkout 95c2bc7
git log -1 && make && ./lz4 --version && ( yes | head -c 65533 | ./lz4 -B4 >file_cout )
# NG

git checkout v1.9.4
git log -1 && make && ./lz4 --version && ( yes | head -c 65533 | ./lz4 -B4 >file_cout )
# NG
```
</details>

**Comment by Cyan4973** [comment]
@MarcMassar : are you able to get `lz4` from the `dev` branch ?
And if yes, could you check if the observed bug is fixed in this branch for you too ?
